lib/bypass/instance.ex

```diff
@@ -79,45 +92,191 @@ defmodule Bypass.Instance do
   defp do_handle_call(:down, _from, %{socket: nil} = state) do
     {:reply, {:error, :already_down}, state}
   end
-  defp do_handle_call(:down, from, %{socket: socket, ref: ref} = state) when not is_nil(socket) do
-    nil = state.caller_awaiting_down  # assertion
-    if state.retained_plug != nil do
-      # wait for the plug to finish
-      {:noreply, %{state | caller_awaiting_down: from}}
+  defp do_handle_call(:down, from, %{socket: socket, ref: ref, callers_awaiting_down: callers_awaiting_down} = state) when not is_nil(socket) do
+    if retained_plugs_count(state) > 0 do
+      # wait for plugs to finish
+      {:noreply, %{state | callers_awaiting_down: [from | callers_awaiting_down]}}
     else
       do_down(ref, socket)
       {:reply, :ok, %{state | socket: nil}}
     end
   end
 
-  defp do_handle_call({:expect, nil}, _from, state) do
-    {:reply, :ok, %{state | expect_fun: nil, request_result: :ok}}
+  defp do_handle_call({expect, fun}, from, state) when expect in [:expect, :expect_once] do
+    do_handle_call({expect, :any, :any, fun}, from, state)
+  end
+
+  defp do_handle_call(
+    {expect, method, path, fun}, _from, %{expectations: expectations} = state)
+      when expect in [:expect, :expect_once]
+      and method in ["GET", "POST", "HEAD", "PUT", "DELETE", "OPTIONS", "CONNECT", :any]
+      and (is_binary(path) or path == :any)
+      and is_function(fun, 1)
+  do
+    route = {method, path}
+
+    updated_expectations =
+      case Map.get(expectations, route, :none) do
+        :none ->
+          Map.put(expectations, route, new_route(
+            fun,
+            case expect do
+              :expect -> :once_or_more
+              :expect_once -> :once
+            end
+          ))
+        _ ->
+          raise ExUnit.AssertionError, "Route already installed for #{method}, #{path}"
+      end
+
+    {:reply, :ok, %{state | expectations: updated_expectations}}
   end
-  defp do_handle_call({:expect, fun}, _from, state) do
-    {:reply, :ok, %{state | expect_fun: fun, request_result: {:error, :not_called}}}
+
+  defp do_handle_call(
+    {expect, _, _, _}, _from, state)
```

Comment:
I am one test away from having tests passing.  Didn't get it all done.  Getting close, though.